### PR TITLE
[L1T] [Phase-2] Update Correlator code

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

This is an update of the correlator code 

Includes:

https://github.com/p2l1pfp/cmssw/pull/92 by @thesps and @slaurila
https://github.com/p2l1pfp/cmssw/pull/94 by @thomreis and @cerminar (only adds a new feature to the emulator, doesn't change any of the existing emulated values)
https://github.com/p2l1pfp/cmssw/pull/96 by @thesps
https://github.com/p2l1pfp/cmssw/pull/97 by @drankincms
https://github.com/p2l1pfp/cmssw/pull/98 by @drankincms (not used by default)
https://github.com/p2l1pfp/cmssw/pull/100 by @thesps
https://github.com/p2l1pfp/cmssw/pull/101 by @gpetruc (not used in production)
https://github.com/p2l1pfp/cmssw/pull/105 by @gpetruc (not used in production)
https://github.com/p2l1pfp/cmssw/pull/107 by @gpetruc (transparent)

Local: https://github.com/cms-l1t-offline/cmssw/pull/1063

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
Hi @cecilecaillol 
I am wondering if we usually add .root or .pb files in CMSSW, or we should make an external. I see that you add
+ L1Trigger/Phase2L1ParticleFlow/data/jecs/jecs_20220308.root
+ L1Trigger/Phase2L1ParticleFlow/data/modelTT_PUP_Off_dXY_XYCut_Graph.pb
in this PR. @smuzaffar @makortel could you please comment on this, thanks.

**Comment by makortel** [comment]
I agree, the `.root` and `.pb` files of this PR should be placed in a data external package (currently this PR would be adding ~900 kB to the repository), and the commits should be cleaned such that the these files do not become part of the commit history.

File: L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+         const std::string &iWeightFile,
+         int iNParticles);
+  ~BJetId();
+
+  void setNNVectorVar();
+  float EvaluateNN();
+  float compute(const l1t::PFJet &iJet, float vz, bool useRawPt);
+
+private:
+  tensorflow::Session *session_;
```
Comment (makortel) [inline L1Trigger/Phase2L1ParticleFlow/interface/BJetId.h:28]: Note that Tensorflow `Session` is thread safe, and would be better placed into the `BJetTFCache`. Nowadays the `tensorflow::run()` accepts also `tensorflow::Session const*`.

**Comment by srimanob** [comment]
> I agree, the `.root` and `.pb` files of this PR should be placed in a data external package (currently this PR would be adding ~900 kB to the repository), and the commits should be cleaned such that the these files do not become part of the commit history.

Hi @makortel 
Thanks. By cleaning, do you mean merge commits?

**Comment by makortel** [comment]
> By cleaning, do you mean merge commits?

Squashing commits together is one option. An alternative is to edit the commits such that the files get removed from them.

**Comment by srimanob** [comment]
Hi @cecilecaillol 
Do you manage to follow @makortel comment on commits
https://github.com/cms-sw/cmssw/pull/40670#issuecomment-1450265835

**Comment by aloeliger** [comment]
@cecilecaillol any update on this?

**Comment by cecilecaillol** [comment]
> > By cleaning, do you mean merge commits?
> 
> Squashing commits together is one option. An alternative is to edit the commits such that the files get removed from them.

I do not want to squash the commits because I would like to keep the history and the original authors as it is a big PR involving work from many people over a long period of time. Could you please let me know how to edit the commit in case the current PR is not ok as it is now?

**Comment by makortel** [comment]
> > > By cleaning, do you mean merge commits?
> > 
> > 
> > Squashing commits together is one option. An alternative is to edit the commits such that the files get removed from them.
> 
> I do not want to squash the commits because I would like to keep the history and the original authors as it is a big PR involving work from many people over a long period of time. Could you please let me know how to edit the commit in case the current PR is not ok as it is now?

Here is one recipe https://www.baeldung.com/git-remove-file-commit-history I found (but didn't try myself).

**Comment by aloeliger** [comment]
@cecilecaillol Is there a status update on this? Jet and correlator PRs are beginning to notice conflicts with central CMSSW when trying to open phase-2 upgrade PRs because they are relying on content here. Would you prefer I take this PR over?

**Pull request closed without merging**